generator_gomod.go

```diff
@@ -45,26 +53,120 @@ func withGomod(g *SourceGenerator, srcSt, worker llb.State, opts ...llb.Constrai
 				// First download the go module deps to our persistent cache
 				// Then set the GOPROXY to the cache dir so that we can extract just the deps we need
 				// This allows us to persist the module cache across builds and avoid downloading
 				// the same deps over and over again.
-				ShArgs(`set -e; GOMODCACHE="${TMP_GOMODCACHE}" go mod download; GOPROXY="file://${TMP_GOMODCACHE}/cache/download" go mod download`),
-				llb.IgnoreCache,
+				ShArgs(`set -e; GOMODCACHE="${TMP_GOMODCACHE}" `+scriptPath+`; GOPROXY="file://${TMP_GOMODCACHE}/cache/download" `+scriptPath),
+				g.withGomodSecretsAndSockets(),
+				llb.AddMount(scriptMountpoint, script),
 				llb.AddEnv("GOPATH", "/go"),
+				credHelper,
 				llb.AddEnv("TMP_GOMODCACHE", proxyPath),
+				llb.AddEnv("GIT_SSH_COMMAND", "ssh -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no"),
 				llb.Dir(filepath.Join(joinedWorkDir, path)),
 				srcMount,
 				llb.AddMount(proxyPath, llb.Scratch(), llb.AsPersistentCacheDir(GomodCacheKey, llb.CacheMountShared)),
 				WithConstraints(opts...),
 			).AddMount(gomodCacheDir, in)
 		}
+
 		return in
 	}
 }
 
+func (g *SourceGenerator) gitconfigGeneratorScript(scriptPath string) llb.State {
+	var script bytes.Buffer
+
+	sortedHosts := SortMapKeys(g.Gomod.Auth)
+	if len(sortedHosts) > 0 {
+		fmt.Fprintln(&script, `set -eu`)
+		script.WriteRune('\n')
+	}
+
+	goPrivate := []string{}
+
+	for _, host := range sortedHosts {
+		auth := g.Gomod.Auth[host]
+		gpHost, _, _ := strings.Cut(host, ":")
+		goPrivate = append(goPrivate, gpHost)
+
+		script.WriteRune('\n')
+		if sshConfig := auth.SSH; sshConfig != nil {
+			username := "git"
+			if sshConfig.Username != "" {
+				username = sshConfig.Username
+			}
+
+			// By default, go will make a request to git for the source of a
+			// package, and it will specify the remote url as https://<package
+			// name>. Because SSH auth was requested for this host, tell git to
+			// use ssh for upstreams with this host name.
+			fmt.Fprintf(&script, `git config --global url."ssh://%[1]s@%[2]s/".insteadOf https://%[3]s/`, username, host, gpHost)
+			script.WriteRune('\n')
+			continue
+		}
+
+		var kind string
+		switch {
+		case auth.Token != "":
+			kind = "token"
+		case auth.Header != "":
+			kind = "header"
+		default:
+			continue
+		}
+
+		fmt.Fprintf(&script, `git config --global credential."https://%[1]s.helper" "/usr/local/bin/frontend credential-helper --kind=%[2]s"`, host, kind)
+		script.WriteRune('\n')
+	}
+
+	fmt.Fprintf(&script, "go env -w GOPRIVATE=%s", strings.Join(goPrivate, ","))
+	script.WriteRune('\n')
+	fmt.Fprintln(&script, "cat go.mod && go mod download")
+	return llb.Scratch().File(llb.Mkfile(scriptPath, 0o755, script.Bytes()))
+}
+
+func (g *SourceGenerator) withGomodSecretsAndSockets() llb.RunOption {
+	return runOptionFunc(func(ei *llb.ExecInfo) {
+		if g.Gomod == nil {
+			return
+		}
+
+		const basePath = "/run/secrets"
+
+		for host, auth := range g.Gomod.Auth {
```

Comment:
I know we have a `fillDefaults` function that makes sure to initialize the Auth map. Are we 100% sure there is not a code path that will hit this with a nil `Auth` value? Should we add a check just in case?

Reply:
If `Auth` is nil, the loop will not execute:
[https://go.dev/ref/spec#For_range](https://go.dev/ref/spec#For_range) (see #3 in the bullet points, if the map is nil the number of iterations is zero)